toolchain/base/canonical_value_store.h

```diff
@@ -22,7 +22,9 @@ namespace Carbon {
 // `KeyT` can optionally be different from `ValueT`, and if so is used for the
 // argument to `Lookup`. It must be valid to use both `KeyT` and `ValueT` as
 // lookup types in the underlying `Set`.
-template <typename IdT, typename KeyT, typename ValueT = KeyT>
+template <typename IdT, typename KeyT, typename ValueT = KeyT,
+          auto ValueToKeyFn = [](typename ValueStoreTypes<ValueT>::ConstRefType
+                                     value) { return value; }>
```

Comment:
Consider
```suggestion
          auto ValueToKeyFn = [](typename ValueStoreTypes<ValueT>::ConstRefType
                                     value) -> decltype(auto) { return value; }>
```
to return a reference to the value instead of a copy.

Reply:
Going instead with the explicit `typename ValueStoreTypes<ValueT>::ConstRefType`, that's where I'd started anyways.